File: clients/csharp/Client/QueryTypes.cs
using Ecs.Protocol.V1;
using Google.Protobuf;

namespace Engine.Core;

/// <summary>
/// A component type a query touches, and how. Produced by <see cref="Query"/>.
/// </summary>
/// <remarks>
/// This is the bridge between the two notions of typing. Authoring code writes
/// <c>Query.Read&lt;Position&gt;()</c> and gets compile-time safety from the generated
/// C# type; the SDK turns it into <c>{ component, schema_hash, access }</c>, which is
/// all the world ever sees. Nothing needs the world to have compiled <c>Position</c>.
/// </remarks>
public readonly record struct ComponentAccess(
    ComponentTypeRef Type,
    Access Access,
    ComponentTypeDeclaration Declaration)
{
    public string Name => Type.LogicalName;

    public bool IsWrite => Access == Access.Write;
}

/// <summary>
/// Declares component access in a query builder.
/// </summary>
public static class Query
{
    /// <summary>Read-only access. Two systems reading the same type run in parallel.</summary>
    public static ComponentAccess Read<T>() where T : IMessage<T>, new() =>
        new(ComponentType<T>.Ref, Access.Read, ComponentType<T>.Declaration);

    /// <summary>
    /// Read-write access. Conflicts with any other system reading or writing the type,
    /// so the scheduler puts them in different stages.
    /// </summary>
    public static ComponentAccess Write<T>() where T : IMessage<T>, new() =>
        new(ComponentType<T>.Ref, Access.Write, ComponentType<T>.Declaration);
}

/// <summary>
/// The dense type ids the coordinator bound this process's schemas to.
/// </summary>
/// <remarks>
/// Ids are assigned by the world, not chosen by the client, so they are only meaningful
/// after the registration handshake. Everything on the wire after that point — queries,
/// batches, leases — is expressed in them.
/// </remarks>
public sealed class SchemaBindings
{
    private readonly Dictionary<string, uint> _byName = new(StringComparer.Ordinal);
    private readonly Dictionary<uint, string> _byId = new();

    public void Add(string logicalName, uint typeId)
    {
        _byName[logicalName] = typeId;
        _byId[typeId] = logicalName;
    }

    public bool TryGetId(string logicalName, out uint typeId) =>
        _byName.TryGetValue(logicalName, out typeId);

    public uint Require(string logicalName) =>
        _byName.TryGetValue(logicalName, out var typeId)
            ? typeId
            : throw new InvalidOperationException(
                $"Component type '{logicalName}' has no id — its schema was never registered. " +
                "Declare it in a query, or add a component of that type through a command buffer.");

    public uint Require<T>() where T : IMessage<T>, new() => Require(ComponentType<T>.Name);

    public string NameOf(uint typeId) =>
        _byId.TryGetValue(typeId, out var name) ? name : $"<unbound:{typeId}>";

    public int Count => _byName.Count;
}

File: clients/typescript/src/query-types.ts
import type { DescMessage } from "@bufbuild/protobuf";
import { Access } from "@ecs/protos/ecs/protocol/v1/query_pb.js";
import { componentType, type ComponentType } from "./component-type.js";

/**
 * A component type a query touches, and how.
 *
 * The bridge between the two notions of typing: authoring code writes
 * `Query.read(PositionSchema)` and keeps compile-time safety from the generated type,
 * while the wire carries only `{ type_id, access }`. Nothing requires the world to have
 * compiled `Position`.
 */
export interface ComponentAccess<Desc extends DescMessage = DescMessage> {
  readonly type: ComponentType<Desc>;
  readonly access: Access;
}

export const Query = {
  /** Read-only. Two systems reading the same type run in parallel. */
  read<Desc extends DescMessage>(schema: Desc): ComponentAccess<Desc> {
    return { type: componentType(schema), access: Access.READ };
  },

  /**
   * Read-write. Conflicts with any other system reading or writing the type, so the
   * scheduler puts them in different stages.
   */
  write<Desc extends DescMessage>(schema: Desc): ComponentAccess<Desc> {
    return { type: componentType(schema), access: Access.WRITE };
  },
} as const;

/**
 * The dense type ids the coordinator bound this process's schemas to.
 *
 * Ids are assigned by the world, so they are meaningless until the registration
 * handshake returns. Everything after that point is expressed in them.
 */
export class SchemaBindings {
  private readonly byName = new Map<string, number>();
  private readonly byId = new Map<number, string>();

  add(logicalName: string, typeId: number): void {
    this.byName.set(logicalName, typeId);
    this.byId.set(typeId, logicalName);
  }

  tryGetId(logicalName: string): number | undefined {
    return this.byName.get(logicalName);
  }

  require(logicalName: string): number {
    const typeId = this.byName.get(logicalName);
    if (typeId === undefined) {
      throw new Error(
        `Component type '${logicalName}' has no id — its schema was never registered. ` +
          "Declare it in a query, or add a component of that type through a command buffer.",
      );
    }
    return typeId;
  }

  nameOf(typeId: number): string {
    return this.byId.get(typeId) ?? `<unbound:${typeId}>`;
  }

  get size(): number {
    return this.byName.size;
  }
}
